File: src/core/store.ts
import * as vscode from "vscode";
import { Reference } from "./reference";

export abstract class Store<T extends string = string> {
  abstract pattern: string;
  protected abstract store: Map<string, Reference[]>;
  abstract parse(uri: vscode.Uri): Promise<void> | void;
  get(key: T): Reference[] {
    return this.store.get(key) ?? [];
  }
  getFrom(uri: vscode.Uri, key: T) {
    return (this.store.get(key) ?? []).filter((v) => v.uri.fsPath === uri.fsPath);
  }
  delete(uri: vscode.Uri): void {
    for (const [id, values] of this.store) {
      const filtered = values.filter((v) => v.uri.fsPath !== uri.fsPath);
      this.store.set(id, filtered);
    }
  }
}

File: rockide/core/store.go
package core

import (
	"github.com/rockide/protocol"
	"go.lsp.dev/uri"
)

type Reference struct {
	Value string
	Uri   uri.URI
	Range *protocol.Range
}

type Store interface {
	Parse(uri uri.URI) error
	Get(key string) []Reference
	GetFrom(uri uri.URI, key Store) []Reference
	Delete(uri uri.URI)
	GetPattern() string
}

func Difference(a []Reference, b []Reference) []Reference {
	result := []Reference{}
	set := map[string]bool{}
	for _, ref := range b {
		set[ref.Value] = true
	}
	for _, ref := range a {
		if !set[ref.Value] {
			result = append(result, ref)
		}
	}
	return result
}
